services/api/service.go

```diff
@@ -484,94 +486,134 @@ func (api *RelayAPI) handleRegisterValidator(w http.ResponseWriter, req *http.Re
 		"mevBoostV": common.GetMevBoostVersionFromUserAgent(ua),
 	})
 
-	respondError := func(code int, msg string) {
-		log.Warn("bad request: ", msg)
-		api.RespondError(w, code, msg)
-	}
-
 	start := time.Now()
 	registrationTimeUpperBound := start.Add(10 * time.Second)
 
-	registrations := []types.SignedValidatorRegistration{}
+	numRegTotal := 0
+	numRegProcessed := 0
+	numRegActive := 0
 	numRegNew := 0
+	processingStoppedByError := false
 
-	if err := json.NewDecoder(req.Body).Decode(&registrations); err != nil {
-		respondError(http.StatusBadRequest, "failed to decode payload")
+	respondError := func(code int, msg string) {
+		processingStoppedByError = true
+		log.Warnf("error: %s", msg)
+		api.RespondError(w, code, msg)
+	}
+
+	body, err := io.ReadAll(req.Body)
+	if err != nil {
+		respondError(http.StatusBadRequest, "failed to read request body")
 		return
 	}
 
-	// Possible optimisations:
-	// - GetValidatorRegistrationTimestamp could keep a cache in memory for some time and check memory first before going to Redis
-	// - Do multiple loops and filter down set of registrations, and batch checks for all registrations instead of locking for each individually:
-	//   (1) sanity checks, (2) IsKnownValidator, (3) CheckTimestamp, (4) Batch SetValidatorRegistration
-	for _, registration := range registrations {
-		if registration.Message == nil {
-			respondError(http.StatusBadRequest, "registration without message")
+	_, err = jsonparser.ArrayEach(body, func(value []byte, dataType jsonparser.ValueType, offset int, _err error) {
```

Comment:
I'd really love to see the function being elsewhere. If you need to capture arguments, you can still call a function that returns the parsing function with explicit captures.

Reply:
fixed after our call to make it more functional